src/Bicep.Core/Configuration/ConfigurationManager.cs

```diff
@@ -0,0 +1,106 @@
+
+        public RootConfiguration GetConfiguration(Uri sourceFileUri)
+        {
+            var customConfigurationPath = DiscoverConfigurationFile(fileSystem.Path.GetDirectoryName(sourceFileUri.LocalPath));
+
+            if (customConfigurationPath is not null)
+            {
+                IConfiguration? customConfiguration;
+
+                try
+                {
+                    using var stream = fileSystem.FileStream.Create(customConfigurationPath, FileMode.Open, FileAccess.Read);
+                    var builder = new ConfigurationBuilder()
+                        .AddInMemoryCollection(builtInRawConfiguration.AsEnumerable())
+                        .AddJsonStream(stream);
+
+                    customConfiguration = builder.Build();
+                }
+                catch (JsonException exception)
+                {
+                    throw new ConfigurationException($"Failed to parse the contents of the Bicep configuration file \"{customConfigurationPath}\" as valid JSON: \"{exception.Message}\".");
+                }
+                catch (Exception exception) when (exception is IOException or UnauthorizedAccessException or SecurityException)
+                {
+                    throw new ConfigurationException($"Could not load the Bicep configuration file \"{customConfigurationPath}\": \"{exception.Message}\".");
+                }
+
+                return RootConfiguration.Bind(customConfiguration, customConfigurationPath);
+            }
+
+            return GetBuiltInConfiguration();
+        }
+
+        private static IConfiguration BuildBuiltInRawConfiguration()
+        {
+            using var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(BuiltInConfigurationResourceName);
+            Debug.Assert(stream is not null, "Default configuration file should exist as embedded resource.");
+
+            var builder = new ConfigurationBuilder();
+            builder.AddJsonStream(stream);
+
+            return builder.Build();
+        }
+
+        private string? DiscoverConfigurationFile(string? currentDirectory)
+        {
+            try
+            {
+                while (!string.IsNullOrEmpty(currentDirectory))
+                {
+                    var configurationPath = this.fileSystem.Path.Combine(currentDirectory, LanguageConstants.BicepConfigurationFileName);
+
+                    if (this.fileSystem.File.Exists(configurationPath))
+                    {
+                        return configurationPath;
+                    }
+
+                    currentDirectory = this.fileSystem.Directory.GetParent(currentDirectory)?.FullName;
+                }
+            }
+            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException or SecurityException)
+            {
+                throw new ConfigurationException($"Error while discovering Bicep configuration file: \"{exception.Message}\".");
```

Comment:
I don't think I expressed this clearly in my last comment on this method, but I'd be a little worried here that there are legitimate scenarios where this could happen with unusual folder configurations (for example, revoking user permissions on a parent folder) - in which case we'd block compilation entirely.  Is that a risk here?

Reply:
Ahh gotcha. That is a risk. I get what you meant by false positives now. For IO related exceptions it makes sense to add telemetry here and return null. I think we probably still need to throw for other exceptions as they would indicate code defect.

Reply:
Raised https://github.com/Azure/bicep/issues/4641 to track the task for adding telemetry and changed it back to return `null`. Also added comments to explain why we don't want to throw when there are IO exceptions.

Reply:
Thanks, sounds good!